### ISSUE TITLE
Update mkl-dnn/OneDNN to 3.2

### ISSUE THREAD

**Issue by anthony-linaro** [OpenImageDenoise/oidn#168]
I am working on porting OIDN to Windows ARM64 platforms as a part of my work on Blender, and as part of that, I have merged some changes to OneDNN upstream: https://github.com/oneapi-src/oneDNN/pull/1654

Would it be possible to update the base of the forked version you use in OIDN2 to 3.2? I would submit a PR myself, but I can't easily work out what changes you have made, as you diverged a while ago.

As an FYI, 3.2 also contains some fixes applicable to MacOS also:
* https://github.com/oneapi-src/oneDNN/pull/1648 where registers were clobbered inadvertently.

**Comment by atafra** [comment]
OIDN currently does not have generic ARM support, only on macOS. It does not use oneDNN at all on macOS ARM, so those fixes are not needed for now. We could upgrade to oneDNN 3.2 but currently this is not a priority because of this.

**Comment by anthony-linaro** [comment]
I am happy to contribute ARM support for Windows! It just relies on OneDNN being updated :)

**Comment by atafra** [comment]
Thanks! Have you already looked into the details of how to add generic ARM support? Upgrading to oneDNN won't be enough.

**Comment by anthony-linaro** [comment]
Using this as a basis, it seems possible: https://github.com/OpenImageDenoise/oidn/pull/134/files

When you say generic ARM support, are you anticipating Linux ARM64 to be added alongside also?

**Comment by atafra** [comment]
It's surely possible but it's not quite as simple as in that PR. That PR replaces our oneDNN fork with the official one which is not an option for us due to the huge and unnecessary binary size increase. So ARM support needs to be added using the forked version but this means that building ARM support in oneDNN needs to be manually implemented in OIDN for the fork. I haven't looked into how oneDNN adds ARM support. If it adds a third-party dependency, that makes things a lot more complicated. Another challenge is that it's not enough to add compilation support to ARM but this also needs to be validated and supported. This is a non-trivial amount of work. And yes, if generic ARM support is added, both Windows and Linux should be supported.

**Comment by anthony-linaro** [comment]
If it's just a case of enabling the code (rather than having to cherry-pick it), then that makes things easier!

Are there any details/is there any documentation of modifications that have been made?

For the steps:
* xbyak_aarch64 is the only dependency that is mandatory - this is actually pulled inline into the OneDNN tree, so is present already. Arm CL is an optional dependency, but doesn't work on ARM64 Windows regardless (but still under MIT, so compatible).
* Is this just a case of enable bits one by one until the build errors go away? Is there a test suite of some sort that can ensure I have the coverage required?
* CMAKE file modification is possible, not a problem - I would guess the above point goes hand in hand with this
* I'm no ML engineer, so pardon if this is a silly question, but is there a test suite I can run for this point?

**Comment by atafra** [comment]
There is no documentation on the changes in the oneDNN fork, sorry.

- I'm wondering why is ArmCL a dependency, even if it's optional. Maybe the default implementation is not performant enough? This also needs to be investigated.
- I'm not sure that enabling bits one by one is a practical solution. Even if you won't get compile errors, you could still have many runtime issues.
- Making the necessary changes in OIDN would require some machine learning knowledge.
- For testing, you could try `oidnTest` as a start.

**Comment by anthony-linaro** [comment]
* ArmCL appears to have been added initially by ARM to provide a framework to speed things up - I am currently running benchDNN to find out the benefits, but initial work can be done without it, if required - regardless it doesn't work on Windows.
* Understood, I can try and do it in one go, the test suites can guide me in that case

For the rest, I can certainly try :) I can call on some ML engineers internally if needed, but ideally I would not have to, as they have their own projects.

Ultimately though, for me to start, ideally I'd need OneDNN to be v3.2, but if you don't want to spend the effort, I can make some small changes locally, or alternatively make a PR to your fork with the required fixes.

**Comment by anthony-linaro** [comment]
So as a quick comparison, using the command line (a random one I stole from the benchdnn docs, not sure which operation reflects OIDN the closest):
```
./benchdnn --conv --mode=p --dt=f32 --dir=FWD_B --attr-post-ops=relu --batch=inputs/conv/set_conv_all
```
Gives Non-ArmCL:
```
total perf: min(ms):55365 avg(ms):61009.9
```
and Enabled-ArmCL:
```
total perf: min(ms):50695.6 avg(ms):55377.2
```

For a ~10% improvement, using `gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0`

**Comment by atafra** [comment]
Thanks. Using ArmCL is not worth the trouble with such a small performance improvement.

**Comment by atafra** [comment]
oneDNN isn't used by default anymore, so I'm closing this. ARM64 support for Windows and Linux is now available starting with `v2.2.0`.
